pkg/auth/providers/saml/saml_client.go

```diff
@@ -0,0 +1,385 @@
+	EntityID          string                  `xml:"entityID,attr"`
+	IDPSSODescriptors []saml.IDPSSODescriptor `xml:"IDPSSODescriptor"`
+	SPSSODescriptors  []saml.SPSSODescriptor  `xml:"SPSSODescriptor"`
+}
+
+const (
+	redirectBackBase = "redirectBackBase"
+	redirectBackPath = "redirectBackPath"
+)
+
+var Root *mux.Router
+
+func InitializeSamlServiceProvider(configToSet *v3.SamlConfig, name string) error {
+
+	var idpURL string
+	var privKey *rsa.PrivateKey
+	var cert *x509.Certificate
+	var err error
+	var ok bool
+
+	if configToSet.IDPMetadataURL == "" {
+		idpURL = ""
+		if configToSet.IDPMetadataContent == "" {
+			if configToSet.IDPMetadataFilePath == "" {
+				log.Debugf("SAML: Cannot initialize saml SP properly, missing IDP URL/metadata in the config %v", configToSet)
+			}
+		}
+	} else {
+		idpURL = configToSet.IDPMetadataURL
+	}
+
+	if configToSet.SPSelfSignedCert == "" {
+		if configToSet.SPSelfSignedCertFilePath != "" {
+			cert, err := ioutil.ReadFile(configToSet.SPSelfSignedCertFilePath)
+			if err != nil {
+				log.Errorf("SAML: Cannot initialize saml SP, cannot read SPSelfSignedCert file in the config %v, error %v", configToSet, err)
+			}
+			configToSet.SPSelfSignedCert = string(cert)
+		} else {
+			log.Debugf("SAML: Cannot initialize saml SP properly, missing SPSelfSignedCert in the config %v", configToSet)
+		}
+	}
+
+	if configToSet.SPSelfSignedKey == "" {
+		if configToSet.SPSelfSignedKeyFilePath != "" {
+			key, err := ioutil.ReadFile(configToSet.SPSelfSignedKeyFilePath)
+			if err != nil {
+				return fmt.Errorf("SAML: Cannot initialize saml SP, cannot read SPSelfSignedKey file in the config %v, error %v", configToSet, err)
+			}
+			configToSet.SPSelfSignedKey = string(key)
+		} else {
+			log.Debugf("SAML: Cannot initialize saml SP properly, missing SPSelfSignedKey in the config %v", configToSet)
+		}
+	}
+
+	if configToSet.SPSelfSignedKey != "" {
+		// used from ssh.ParseRawPrivateKey
+
+		block, _ := pem.Decode([]byte(configToSet.SPSelfSignedKey))
+		if block == nil {
+			return fmt.Errorf("SAML: no key found")
+		}
+
+		if strings.Contains(block.Headers["Proc-Type"], "ENCRYPTED") {
+			return fmt.Errorf("SAML: cannot decode encrypted private keys")
+		}
+
+		switch block.Type {
+		case "RSA PRIVATE KEY":
+			privKey, err = x509.ParsePKCS1PrivateKey(block.Bytes)
+			if err != nil {
+				return fmt.Errorf("SAML: error parsing PKCS1 RSA key: %v", err)
+			}
+		case "PRIVATE KEY":
+			pk, err := x509.ParsePKCS8PrivateKey(block.Bytes)
+			if err != nil {
+				return fmt.Errorf("SAML: error parsing PKCS8 RSA key: %v", err)
+			}
+			privKey, ok = pk.(*rsa.PrivateKey)
+			if !ok {
+				return fmt.Errorf("SAML: unable to get rsa key")
+			}
+		default:
+			return fmt.Errorf("SAML: unsupported key type %q", block.Type)
+		}
+	}
+
+	if configToSet.SPSelfSignedCert != "" {
+		block, _ := pem.Decode([]byte(configToSet.SPSelfSignedCert))
+		if block == nil {
+			panic("SAML: failed to parse PEM block containing the private key")
```

Comment:
no panics in the code. return errors and ensure it is handled properly higher up. same on line 133